enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,290 @@
+
+#### Declarative container native API
+
+As an OpenShift administrator using a Baremetal platform, I want to have a declarative container native API for network interface configuration rather than have to put ifcfg/NM keyfiles.
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+### Risks and Mitigations
+
+#### Not rebooting could hide errors in network configuration
+
+A NetworkManager profile is configured with certain settings (like connection.autoconnect=yes). See those settings (the content of the profiles) with `nmcli connection show "$PROFILE"`. The profile is persisted to disk (in the file you see in `nmcli -f all connection`). When NM starts, it loads the profiles from disk. NetworkManager will try to automatically connect/activate profiles that are configured to autoconnect.
+
+There is no guarantee that a reboot will succeed to connect to the network. e.g. the DHCP might simply not reply. But in general, if you have a profile (persisted to disk), and the profile is set to autoconnect, and you reboot, there is a reasonable expectation that it should work.
+
+If it doesn't come up on reboot, it's likely an environment issue (e.g. cable unplugged) or a real bug (that will be fixed by NetworkManager team).
+
+## Design Details
+
+#### Option C
+
+These configuration files would be handled via MCO, for the Baremetal platform only.
+
+In /etc/NetworkManager/conf.d/99-kni.conf we add:
+[keyfile]
+path=/etc/NetworkManager/system-connections-merged
+
+In /etc/tmpfiles.d add a file to specify:
+D /tmp/nm-system-connections 0755 root root
+D /tmp/nm-system-connections-work 0755 root root
+
+Create a .mount systemd unit to mount the overlay:
+/etc/NetworkManager/system-connections-merged
+overlay lowerdir=/etc/NetworkManager/system-connections,upperdir=/tmp/nm-system-connections,workdir=/tmp/nm-system-connections-work
+
+All the changes that nmstate will do will end up affecting just /tmp/nm-system-connections so once the machine reboots those will be gone, and only what's on /etc/NetworkManager/system-connections , which is where people would put their keyfiles with MachineConfig would remain.
+
+When kubelet would start, it would start kubernetes-nmstate again, we'd see that NNS doesn't match the NNCP that we need to apply again.
+
+The changes that knmstate + nmstate will do will happen without reboot and since they only get written to the tmpfs won't affect the next boot.
+If you need persistent changes, you do them with MCO.
+Kubernetes-nmstate will only affect post-kubelet so we comply with the "MC owns every configuration until boot time".
+
+[Link](https://asciinema.org/a/uMqwIpvfhuI67ShT12csYXm3h) to an asciicast showing this option.
```

Comment:
This video is really really helpful

Reply:
Yes, you have it correct. We skip the reboot, and the configuration for kubernetes-nmstate is ephemeral on disk.

Reply:
awesome thanks!